lib/auth.js

```diff
@@ -3,25 +3,29 @@
  * @since   1/14/14
  */
 
-(function() {
+(function () {
 
     var _ = require('lodash')
         , util = require('util')
         , request = require('request')
         , oauth = require('oauth')
         , qs = require('querystring')
-        , stateIn;
+        , stateIn = [];
 
-    var Auth = function(Inherits, args) {
+    var Auth = function (Inherits, args) {
         Inherits(this);
 
-        stateIn = args.state || Math.random().toString();
+        if (args.state && args.state.constructor === Array)
+            stateIn = args.state;
+        else
+            stateIn.push(args.state || Math.random().toString());
+        this.hasCustomState = args.state ? true : false;
 
         this.options = {};
 
-        this.authorize = function(res, scope) {
+        this.authorize = function (res, scope, state) {
 
-            if(Object.prototype.toString.call(res) == "[object Array]"){
+            if (Object.prototype.toString.call(res) == "[object Array]") {
```

Comment:
I think you should better change it into:

```javascript
if (res instanceof Array) {
   ....
}
```

That's more prominent and cleaner way :)